### ISSUE TITLE
"What's New" link inside the menu doesn't open in a modal view

### ISSUE THREAD

**Issue by brampitoyo** [mozilla-mobile/focus-android#2107]
Steps to reproduce:

- Tap `menu` -> `Help`. Notice that the link opens in an in-content modal view. This enables you to go back by tapping the arrow on the top-left corner.
  - Try it with other links, e.g. `menu` -> `Settings` -> `Privacy Notice`
- Tap `menu` -> `What’s new`. Notice that the link opens in a Focus app tab.

The behaviour of “What’s New” should follow the behaviour of other in-content links.

**Comment by pocmo** [comment]
I would prefer the opposite: Open tabs for all of them. Back when we implemented "Help" etc. we didn't support multiple tabs yet, so we had to build those separate screens. But they are a pain to maintain.

**Comment by brampitoyo** [comment]
While I think that there are many benefits to be gained from for clearly separating “Pages that belong to the product” from “Every other pages on the web”, I’m really interested in knowing why you wrote that they’re a pain to maintain.

Does it have to do with the fact that these pages had to use a pseudo-tab workaround that had since been obsoleted by the multi-tab feature? I wonder if there’s a way to reimplement modal pages that won’t cause a lot of maintenance woes.

**Comment by pocmo** [comment]
> While I think that there are many benefits to be gained from for clearly separating “Pages that belong to the product” from “Every other pages on the web”, I’m really interested in knowing why you wrote that they’re a pain to maintain.

Because they actually render websites and therefore we need to perform tracking protection there too and potentially break things. Also we can't fully isolate WebViews. Data from websites (Cookies, cache) that the user browses in those screens might affects websites loaded in other tabs. And finally in those screens we are missing all the other controls to reload and navigate back/forward which makes this awkward to use if you click a couple of links.

**Comment by pocmo** [comment]
An example of special casing things because of those modal views is: #1964

**Comment by brampitoyo** [comment]
> Because they actually render websites and therefore we need to perform tracking protection there too
>
> Also we can't fully isolate WebViews. Data from websites (Cookies, cache) that the user browses in those screens might affects websites loaded in other tabs.
>
> And finally in those screens we are missing all the other controls to reload and navigate back/forward

Initially, I was worried about the confusion that may result from presenting in-content pages in the same manner as the rest of the web. But you’ve outlined a few technical as well as UI problems that appear as a result of our current way of doing things, and I agree with all of them.

So then, can we open in-content pages in normal browser tabs without causing confusion?

Yes. I think so, but we have to change our current behaviour a little bit:

1. If there’s no page currently opened, in-content pages may open in the current tab. But **if there’s a page already opened, in-content pages should always open in a new tab**. This way, they will never take over the page content that’s currently opened.
2. Every in-content pages are opened from within Settings (except for “What’s New” on Android, which is directly under the menu). **When an in-content page is opened, the Settings page should animate as if it’s closing to make way for a newly opened tab**. This way, user will understand the lack of Back/Close button on top, and that tapping the hardware Back button wouldn’t return them to the menu.

What do you think?

(If this is a good behaviour to adopt, then I think we can rename this issue, or close it and file a new one).

**Comment by pocmo** [comment]
Yeah, this sounds good! Let's try that!

**Comment by brampitoyo** [comment]
@pocmo Hope this isn’t overthinking, but I’m still cautious about the burden we put on the user by having them manage tabs when all they want to do is get support.

Let’s try this idea first and see whether we can offset this tab management hassle by having clear animations and visual cues.

The other problem I see is that, once the Help page opens in a new tab, there’s no way to close only that specific tab. The only thing the user can do is erasing all the tabs using the garbage bin icon.

**Comment by pocmo** [comment]
> @pocmo Hope this isn’t overthinking, but I’m still cautious about the burden we put on the user by having them manage tabs when all they want to do is get support.

Yeah, I agree. In Fennec we also have some "settings" that instead close the screen and open a tab. And yes, there have been discussions whether this might be confusing.

> The other problem I see is that, once the Help page opens in a new tab, there’s no way to close only that specific tab. The only thing the user can do is erasing all the tabs using the garbage bin icon.

Pressing the hardware back button should close the tab if there's no history to go back to. But it's not that obvious.

**Comment by brampitoyo** [comment]
Posting a draft visual + animation that will help clarify how in-content page can open.

![android in content pages](https://user-images.githubusercontent.com/51007/36958000-6cf34d8c-209d-11e8-8060-e1fa99c94d8b.gif)

As mentioned by @pocmo, it’s not that obvious what pressing the hardware back button will do. On the first press, it’s a simple answer: the tab will close. But on the second press, should we go back to Settings even though the sheet ‘moves forward’ spatially? This may be confusing.

Therefore, rather than committing to this solution, we should try and see how it feels first.

**Comment by pocmo** [comment]
> But on the second press, should we go back to Settings even though the sheet ‘moves forward’ spatially?

This might be confusing too considering that a lot can happen until the user presses back again - and then they might not expect going to settings again. However I'm more concerned here that this will again add engineering overhead because we would need to track the journey the user has been taking through the app.

**Comment by brampitoyo** [comment]
@pocmo You’re right. For now, let’s not worry about the second press of the “Back” button, and focus on designing a good experience for opening in-content links in normal tabs

**Comment by ekager** [comment]
Hey @brampitoyo . I did a quick hack on this and removed InfoActivity to see what it feels like. (no animations). I'll send you a build now.
https://github.com/ekager/focus-android/tree/tab_pages

**Comment by brampitoyo** [comment]
Hi @pocmo and @ekager, I just had a look at the `tab_pages` build. This is the start of a good approach! My observations below.

1. When tapping “What’s New” and “Help” from within the hamburger menu, it already feels as if these links should open in a browser tab. The things we will need here are:

* Animate the transition of the URL bar colour from its initial state (saturated) to its loading state (desaturated)

<img src="https://user-images.githubusercontent.com/51007/37572214-aebb2738-2b6c-11e8-92f6-a6b6261ba901.gif" height="80">

* Animate the FAB button from its initial state (nothing) to showing a garbage bin icon, as per [Material Design Guidelines]()

![floating-action-button-animate-in](https://user-images.githubusercontent.com/51007/37572505-dc0c6a36-2b70-11e8-9492-915deb2982d1.gif)

2. When tapping in-content links from within the Settings, the transition-out we currently implement (ie. sheet which contains the Settings slides down) would work fine without any modification, because it matches our transition-in (ie. sheet which contains the Settings slides up).
    * I suggest not implementing the slide-right transition I proposed earlier in this issue
    * However, the URL bar colour and FAB button animations on point 1 are still important and necessary

I’m really curious to see if the animations plus link icons will help make our open in new tab interaction look and feel as smooth and seamless as opening in custom tab!

**Comment by ekager** [comment]
After talking to @brampitoyo , I'm going to investigate adding these animation features for this issue and also all of Focus (especially FAB)

**Comment by brampitoyo** [comment]
I’m CCing @amin who will be able to review this work in context of semantic animations – hopefully before his PTO!

Hi Amin, our story goes like this.

### Background

In Focus, we have in-content links both inside our hamburger menu, and inside our settings page. These links are e.g. Help, About Firefox Focus and Your Rights.

Currently, we open these in-content links in a modal tab. Think of it like going one level deeper into another settings page that slides up the view. There’s a back button on this modal view. Tap it, and the modal view slides down and you’re back to Settings.

Unfortunately, we have found that this method of opening in-content links is a pain to maintain – and for good reason. Running an in-app browser tab within a browser that has tabs is bound to get us into trouble! So we’ve decided to open these links in a normal browser tab.

### Problem

According to the user’s mental model, tapping an in-content link should move them one level deeper into the Settings. However, what we’re doing is taking them _out of_ Settings and into the browser view.

### Potential solution

Firstly, we clearly label every in-content link with an `open in new` icon:

<img src="https://user-images.githubusercontent.com/51007/37743195-cc07d274-2dcd-11e8-8b53-238fdfa51685.png" height="44">

Secondly, we animate some page elements. This is to help users understand that:

1. They’re moving back in the stack, and
2. A new tab is opened with the in-content URL in it

#### Open an in-content link when another tab is already open

(Please disregard the slide-right animation. It’s not the effect that Android has. The correct effect is slide down).

![animation](https://user-images.githubusercontent.com/51007/36958000-6cf34d8c-209d-11e8-8060-e1fa99c94d8b.gif)

#### Open an in-content link when no tab is open

The URL bar should fade – same as the example above:

<img src="https://user-images.githubusercontent.com/51007/37572214-aebb2738-2b6c-11e8-92f6-a6b6261ba901.gif" height="80">

Then, because there’s no FAB button on the screen yet, we animate it to appear as per Material Guidelines:

![fab](https://user-images.githubusercontent.com/51007/37572505-dc0c6a36-2b70-11e8-9492-915deb2982d1.gif)

### For you to think about, if you have time

Does this animation help user understand fact that they’re moving back, and that a new tab is opened?

**Comment by aminalhazwani** [comment]
Hey @brampitoyo thanks for reaching out. This looks really interesting! 

For a truly semantic animation we should opt for something like this

<img width="1054" alt="screen shot 2018-03-22 at 1 03 52 pm" src="https://user-images.githubusercontent.com/1721875/37769540-a31f89d4-2dd1-11e8-8eb8-632ecafdfa29.png">

where the tab grows out of the list item. Now, of course I understand that we have some OS constrains that for the time being we have to acknowledge, so there are a couple of things we could add to make this even more clear.

Choreography plays a big role here: there are a lot of things moving. How can we direct the users eyes in order to let them understand what is happening? The FAB button is a good actor, it can really help users understand that a second tab is being open. It would be interesting to explore more how to sync and direct different animations because FAB, progress bar and URL bar in the demo are animated all at the same time. 

Related to the FAB what was initially a Erase button becomes a Tabs button. Some improvements that we could consider: show immediately the "1" and then bump it to "2". We could also use the same tab signifier that we use for tabs on Firefox for Android. 

<img width="71" alt="screen shot 2018-03-22 at 1 06 43 pm" src="https://user-images.githubusercontent.com/1721875/37771186-1c70f278-2dd7-11e8-8287-8bdbb31a3d98.png">

Another possible small improvement is the icon choice for the "Help" list item. While we use the "Open in New" for external links in web properties in this case we could use the "Tab New" icon since everything is happening inside the app and would reinforce the idea that a new tab is being open.

<img width="271" alt="screen shot 2018-03-22 at 1 06 51 pm" src="https://user-images.githubusercontent.com/1721875/37771219-398dee06-2dd7-11e8-93ca-db0d3ea38ba2.png">

I hope it helps, and let me know what do you think and how you will move forward!

**Comment by brampitoyo** [comment]
Thanks for your concrete and thorough feedbacks, @aminalhazwani! I think that all of them ought to be done, but not in version 4.2 as we’re quickly running out of time.

However, your “Tab New” icon can possibly be included in this iteration!

I will copy and paste your feedback into a follow-up issue, noting that it should be done in future versions.
